### PR TITLE
Add VERSIONING document and make sure all internal packages have doc …

### LINKED ISSUE #2654: Document the requirements and guarantees for the internal packages
I would add an extra sentence that this may break at any moment to scare users of using it.

### LINKED ISSUE #2748: Follow the versioning and stability requirements
From the specification this repo MUST define a language-specific document which details how this repository will follow the versioning and stability requirements. This document SHALL be placed in the root of each repo and named VERSIONING.

Probably extract the Release section of the main readme in a `VERSIONING` document.

### PR DESCRIPTION
**Pull request by anuraaga**
…about internalness.

Fixes #2748
Fixes #2654 (let me know if this isn't enough, I'll remove this and follow up)

### CODE REVIEW

**Review by jkwatson — APPROVED** [review]
Thanks!

File: VERSIONING.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+A backwards-incompatible change affects the public API of a module. The public API is any public
+class or method that is not in a package which includes the word `internal`. Examples of incompatible
+changes are:
+
+- API changes that could require code using the artifact to be changed, e.g., removing a method, 
+  reordering parameters, adding a method to an interface or abstract class without adding a default
+  implementation.
+  
+- ABI changes that could require code using the artifact to be recompiled, but not changed, e.g.,
```
Comment (bogdandrutu) [inline VERSIONING.md:21]: This is a great resource https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html

File: VERSIONING.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+Such changes will be avoided - if they must be made, the `MAJOR` version of the artifact will be
+incremented.
+
+Backwards incompatible changes to `internal` packages are expected. Versions of published artifacts
+are expected to be aligned by using BOMs we publish. We will always provide BOMs to allow alignment
+of versions.
+
+As a user, if you always depend on the latest version of the BOM for a given `MAJOR` version, and
+you do not use classes in the `internal` package, you can be assured that your app will always
```
Comment (bogdandrutu) [inline VERSIONING.md:37]: you do not use classes in the `internal` package (which you MUST NOT do)

File: api/context/src/main/java/io/opentelemetry/context/internal/shaded/package-info.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Interfaces and implementations that are internal to OpenTelemetry.
+ *
+ * <p>All the content under this package and its subpackages are considered not part of the public
+ * API, and should not be used by users of the OpenTelemetry library.
```
Comment (bogdandrutu) [inline api/context/src/main/java/io/opentelemetry/context/internal/shaded/package-info.java:5]: s/should not/MUST NOT :)

**Review by carlosalberto — APPROVED** [review]

File: VERSIONING.md (hunk 1 above)
Comment (anuraaga) [reply VERSIONING.md:21]: Thanks, updated

**Review by bogdandrutu — APPROVED** [review]

**Review by iNikem — APPROVED** [review]

**Pull request merged**